### PR TITLE
eclipse-lyo#636 Add option to not add providers by default

### LINKED ISSUE #636: Deprecated Json4J Providers still registered and cannot be prevented
The Providers from the `Json4JProvidersRegistry `are deprecated but still registered by default. This behavior cannot be prevented when using the default OslcClient.
This leads to malformed json when the `OslcRdfJsonProvider `is used as `MessageBodyWriter`.

This issue occurs since lyo 6.0.0 but may have been in there earlier.

Provide an option to prevent registration of providers by default and to add them manually when needed.

### PR DESCRIPTION
**Pull request by jhemm2**
## Description

_Adds another Constructor the OslcClient to allow custom registration of providers for the Client. This is due to registration of deprecated json providers by default_

## Checklist

- [x ] This PR adds an entry to the CHANGELOG. _See https://keepachangelog.com/en/1.0.0/ for instructions. Minor edits are exempt._ (Minor)
- [ ] This PR was tested on at least one Lyo OSLC server (e.g. manual workflow on [Lyo Sample](https://github.com/OSLC/lyo-samples/actions/workflows/maven-smoke-manual.yml) and [OSLC RefImpl](https://github.com/oslc-op/refimpl/actions/workflows/maven-acceptance-manual.yml)) or adds unit/integration tests.
- [ x] This PR does NOT break the API

## Issues

Closes #636 ; 

_(use exactly this syntax to link to issues, one issue per statement only!)_

### CODE REVIEW

**Comment by jhemm2** [comment]
I made a proposal commit and removed the deprecated dependency. Furthermore, I have refactored the OslcClient to avoid duplicate code and added some unit-tests accordingly. Hth

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,7 +9,7 @@
 ### Deprecated
 
 ### Removed
-
+- Dependency to deprecated oslc4j-json4j-provider
```
Comment (berezovskyi) [inline CHANGELOG.md:12]: - Dependency to deprecated `oslc4j-json4j-provider` in Lyo Client

**Review by berezovskyi — APPROVED** [review]

File: CHANGELOG.md (hunk 1 above)
Comment (Jad-el-khoury) [reply CHANGELOG.md:12]: @berezovskyi  is good to clarify/update before merging.

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java [hunk 2]
```diff
@@ -1,14 +1,8 @@
...
- * See the NOTICE file(s) distributed with this work for additional
- * information regarding copyright ownership.
- *
- * This program and the accompanying materials are made available under the
- * terms of the Eclipse Public License 2.0 which is available at
- * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License 1.0
- * which is available at http://www.eclipse.org/org/documents/edl-v10.php.
- *
+ * Copyright (c) 2020 Contributors to the Eclipse Foundation See the NOTICE file(s) distributed with this work for
+ * additional information regarding copyright ownership. This program and the accompanying materials are made available
```
Comment (Jad-el-khoury) [inline client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java:3]: This seems to be an un-intended change. Please revert.

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java [hunk 3]
```diff
@@ -33,577 +29,477 @@
...
-	public Response getResource(String url, Map<String, String> requestHeaders) {
-		return getResource(url, requestHeaders, OSLCConstants.CT_RDF, null, true);
-	}
-
-	public Response getResource (String url, Map<String, String> requestHeaders, String defaultMediaType) {
-	    return getResource(url, requestHeaders, defaultMediaType, null, true);
-	}
-
-    public Response getResource (String url, Map<String, String> requestHeaders, String defaultMediaType,
+    private final String version;
```
Comment (Jad-el-khoury) [inline client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java:53]: The whole file has been changed. 
Main reason is due to white-space changes. But other formatting changes are also introduced.
Please revert such changes unless they are intentional.

File: client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java [hunk 4]
```diff
@@ -33,577 +29,477 @@
...
+    }
+
+    public Response getResource(String url, Map<String, String> requestHeaders, String defaultMediaType,
+            boolean handleRedirects) {
+        return getResource(url, requestHeaders, defaultMediaType, null, handleRedirects);
+    }
+
+    public Response getResource(String url, Map<String, String> requestHeaders, String defaultMediaType,
+            String configurationContext, boolean handleRedirects) {
+        return doRequest(HttpMethod.GET, url, null, configurationContext, null, defaultMediaType, OSLCConstants.CT_RDF,
```
Comment (Jad-el-khoury) [inline client/oslc-client/src/main/java/org/eclipse/lyo/client/OslcClient.java:215]: This is a significant refactoring, and looks promising. But I can't easily review it. Too many other (formatting) changes in the way.
If you are confortable with that @berezovskyi, I am fine.

**Comment by Jad-el-khoury** [comment]
The original change to control the Providers looks good. But noticed other refactoring that I could not review.
Are you happy with them @berezovskyi then I am fine.

but @jhemm2 ! Please fix the unintended changes. I am afraid we will endup with many back-forth changes due to diffferent IDE settings.

**Comment by berezovskyi** [comment]
Yes, I am fine with the changes.

Reformatting the code touched by the PR is intentional and welcomed/required.

With regards to the settings, there is sn .editorconfig file in the repo that everyone (cough) is expected to apply when reformatting.

But I have two further ideas on my bucket list:

- add checkstyle to enforce more than just indentation and line limits.
- do a big bang reformat of the whole codebase - but it will break all pull requests.

**Comment by berezovskyi** [comment]
Two more bits:

- it would be good @jhemm2 if you could commit code changes and reformat as different commits - same PR. Ideally, one commit at the start and one at the end, but not a strict rule.
- @Jad-el-khoury please configure https://github.com/ncjones/editorconfig-eclipse#installation on your machine - as is described in our CONTRIBUTING.md in **bold**: https://github.com/eclipse-lyo/lyo/blob/master/CONTRIBUTING.md#contributing-to-eclipse-lyo

**Comment by Jad-el-khoury** [comment]
@berezovskyi editorconfig seems to be in place, but not sure if it makes a difference. 
go ahead then @jhemm2 !

**Comment by Jad-el-khoury** [comment]
can @jhemm2 commit to master, or should we do it?

**Comment by berezovskyi** [comment]
@Jad-el-khoury not unless @jhemm2 becomes a committer ;)

**Pull request merged**